client/src/app/shared/components/os-sort-filter-bar/os-sort-filter-bar.component.ts

```diff
@@ -0,0 +1,388 @@
+ */
+export interface OsFilterOption {
+    label: string;
+    condition: string | boolean | number;
+    isActive?: boolean;
+}
+
+/**
+ * Reusable bar for list views, offering sorting and filter options.
+ * It will modify the DataSource of the listView to include custom sorting and
+ * filters.
+ *
+ * ## Examples:
+ * ### Usage of the selector:
+ *
+ * ```html
+ * <os-sort-filter-bar model="agenda" [modelRepo]="repo.getViewModelListObservable()"
+ *      [filters]="filterOptions" [sorting]="sortOptions"
+ *      (filteredData)="onFilterData($event)">
+ * </os-sort-filter-bar>
+ * ```
+ * Then, write some filterOptions and sortOptions into the listview.
+ */
+@Component({
+    selector: 'os-sort-filter-bar',
+    templateUrl: './os-sort-filter-bar.component.html',
+    styleUrls: ['./os-sort-filter-bar.component.scss']
+})
+export class OsSortFilterBarComponent<V extends BaseViewModel> implements OnInit {
+
+    @Input()
+    public model: string;
+
+    /**
+     * The definition of the sorting columns to be made available
+     */
+    @Input()
+    public sorting: OsSortingDefinition<V>;
+
+    /**
+     * A list of filters to be made available
+     */
+    @Input()
+    public filters: OsFilter[];
+
+    @ViewChild(MatMenuTrigger)
+    public menuTrigger: MatMenuTrigger;
+
+    /**
+     * The original repository where the data are fetched from
+     */
+    @Input()
+    public modelRepo: Observable<V[]>;
+
+    /**
+     * The filter side drawer
+     */
+    @ViewChild('filterMenu')
+    public filterMenu: FilterMenuComponent;
+
+    /**
+     * The bottom sheet used to alter sorting in mobile view
+     */
+    @ViewChild('sortBottomSheet')
+    public sortBottomSheet: OsSortBottomSheetComponent<V>;
+
+    /**
+     * An event emitter to offer filtered Data to display.
+     */
+    @Output()
+    public filteredData = new EventEmitter<MatTableDataSource<V>>();
+
+    /**
+     * The 'opened/active' state of the fulltext filter input field
+     */
+    public isSearchBar = false;
+
+    /**
+     * Helper data source to assist with MatSort sorting and fulltext filter
+     * @ignore
+     */
+    private _filteredDataSource = new MatTableDataSource<V>();
+
+    /**
+     * The current data that come from the repository
+     */
+    private _originalData: V[];
+
+    public constructor(
+        private bottomSheet: MatBottomSheet,
+        public translate: TranslateService,
+        public vp: ViewportService,
+        private storageService: StorageService
+        ) {
+            this.filterMenu = new FilterMenuComponent(translate);
+            this.filterMenu.activeFilters.subscribe(newFilters => {
+                this.filters = newFilters;
+                this.filterData();
+            });
+    }
+
+    public ngOnInit(): void {
+        this.checkStoredFilterSort();
+        if (this.modelRepo){
+            this.modelRepo.subscribe(originalData => {
```

Comment:
this is why there are performance issues.

1. you don't want to filter everything immediately. You want to start filtering then you decide to to so (applying the first filter)
2. You don't want to sort absolutely everything when there is now data arriving but only the changes.

Reply:
As for (2), we discussed this: The way autoupdate arrives, I see no way avoid the initial sorting problems (sorting one data, sorting two, .... ) right now. I always receive all data, that have no notion of preexisting ones.

Reply:
(1) is obsolete, because the whole filter was changed into a service